### PR TITLE
SAM-1113 Ability to flag question as Extra Credit

### PR DESCRIPTION
**Pull request by davidpbauer**
This allows instructors to flag questions in their assessment as extra credit. Extra credit questions will not count towards the total possible points for an assessment. However, any points earned by the student for answering the extra credit question correctly will be added to  their total earned points for that assessment. Currently, students are able to score over 100 percent on an assessment.

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentBean.java [hunk 1]
```diff
@@ -180,7 +180,8 @@ public void setQuestionSizeAndTotalScore() {
       for (int j=0; j<itemsInThisSection; j++){
           ItemContentsBean item = (ItemContentsBean)items.get(j);
           if (item.getItemData().getScore()!=null){
-            this.totalScore += item.getItemData().getScore().doubleValue();
+            if(item.getItemData().getIsExtraCredit()==null || !item.getItemData().getIsExtraCredit().booleanValue())
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/author/AssessmentBean.java]: Please use braces

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java [hunk 2]
```diff
@@ -711,6 +711,9 @@ public void saveItem(ItemAuthorBean itemauthor) throws FinFormatException{
      		item = new PublishedItemFacade();
      	}
       }
+
+	  item.setIsExtraCredit(bean.getIsExtraCredit());
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java]: Can you check tabs vs spaces here? In an editor like vim, use "set list" to see tab chars

File: samigo/samigo-app/src/webapp/jsf/author/editAssessment.jsp [hunk 3]
```diff
@@ -383,6 +383,8 @@ document.links[newindex].onclick();
 		<h:outputText rendered="#{question.itemData.score > 1}" value=" #{authorMessages.points_lower_case}"/>
 		<h:outputText rendered="#{question.itemData.score == 1}" value=" #{authorMessages.point_lower_case}"/>
 		<h:outputText rendered="#{question.itemData.score == 0}" value=" #{authorMessages.points_lower_case}"/>
+
+        <h:outputText styleClass="extraCreditLabel" rendered="#{question.itemData.isExtraCredit == true}" value=" #{authorMessages.extra_credit_preview}" />
```
Comment (ottenhoff) [inline samigo/samigo-app/src/webapp/jsf/author/editAssessment.jsp]: Please check spaces vs tabs here

File: samigo/samigo-app/src/webapp/jsf/author/item/fillInNumeric.jsp [hunk 4]
```diff
@@ -35,7 +35,7 @@
       <title><h:outputText value="#{authorMessages.item_display_author}"/></title>
       <samigo:script path="/js/authoring.js"/>
       </head>
-      <body onload="countNum();<%= request.getAttribute("html.body.onload") %>">
+      <body onload="countNum();<%= request.getAttribute("html.body.onload") %>">
```
Comment (ottenhoff) [inline samigo/samigo-app/src/webapp/jsf/author/item/fillInNumeric.jsp]: I don't see any changes here

File: samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.java [hunk 5]
```diff
@@ -299,6 +300,10 @@ public void setLastModifiedDate(Date lastModifiedDate) {
     this.lastModifiedDate = lastModifiedDate;
   }
 
+  public Boolean getIsExtraCredit() { return this.isExtraCredit; }
+
+  public void setIsExtraCredit(Boolean isExtraCredit) { this.isExtraCredit = isExtraCredit; }
+
```
Comment (ottenhoff) [inline samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.java]: Please format like other getters/setters in the file

**Comment by davidpbauer** [comment]
Hey Sam,

I've cleaned up the formatting and amended my commit. Do I need to make a new Pull Request or is there a way to "refresh" this one?

David

**Comment by ottenhoff** [comment]
Hi David,

The easy and safe approach: commit your changes to the same SAM-1113 branch on your local and then git push the changes to your origin. The PR will be auto-updated.

The more difficult approach is to rewrite your local history using git rebasing and git force pushing....

File: samigo/samigo-app/src/webapp/jsf/author/item/fillInNumeric.jsp (hunk 4 above)
Comment (davidpbauer) [inline samigo/samigo-app/src/webapp/jsf/author/item/fillInNumeric.jsp]: I couldn't get fillInNumeric.jsp formatted in a way that it wouldn't commit these non-changes. Even VIM thinks the two lines are identical. Could it be something with the line endings? I'm developing on linux.

Comment (ottenhoff) [inline samigo/samigo-app/src/webapp/jsf/author/item/fillInNumeric.jsp]: Yeah, it's prob line endings. Don't worry about it ...

Comment (bjones86) [inline samigo/samigo-app/src/webapp/jsf/author/item/fillInNumeric.jsp]: On Linux I use two utilities to manage line endings: dos2unix and unix2dos

**Comment by jonespm** [comment]
I think this looks good now. I'd have some questions on QTI import/export (if the spec supports it and if this would add this metadata) and also there's the current outstanding question on the grading. Otherwise looks good to merge.

**Comment by raulsv** [comment]
Hi

I've tested the extra credit functionality in the qa server sakai-trunk.atica.um.es and it looks good but I've found some issues. Please see my last comment in the Jira SAM-1113. Thanks

**Comment by steveswinsburg** [comment]
Closing due to inactivity. Please send a new PR if feature/fix is still desired.

**Pull request closed without merging**